jetty-core/jetty-io/src/main/java/org/eclipse/jetty/io/content/BufferedContentSink.java

```diff
@@ -0,0 +1,170 @@
+//
+// ========================================================================
+// Copyright (c) 1995 Mort Bay Consulting Pty Ltd and others.
+//
+// This program and the accompanying materials are made available under the
+// terms of the Eclipse Public License v. 2.0 which is available at
+// https://www.eclipse.org/legal/epl-2.0, or the Apache License, Version 2.0
+// which is available at https://www.apache.org/licenses/LICENSE-2.0.
+//
+// SPDX-License-Identifier: EPL-2.0 OR Apache-2.0
+// ========================================================================
+//
+
+package org.eclipse.jetty.io.content;
+
+import java.io.IOException;
+import java.nio.ByteBuffer;
+
+import org.eclipse.jetty.io.ByteBufferPool;
+import org.eclipse.jetty.io.Content;
+import org.eclipse.jetty.io.RetainableByteBuffer;
+import org.eclipse.jetty.util.BufferUtil;
+import org.eclipse.jetty.util.Callback;
+import org.eclipse.jetty.util.IteratingNestedCallback;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * <p>A {@link Content.Sink} backed by another {@link Content.Sink}.
+ * Any content written to this {@link Content.Sink} is buffered,
+ * then written to the delegate's
+ * {@link Content.Sink#write(boolean, ByteBuffer, Callback)}. </p>
+ */
+public class BufferedContentSink implements Content.Sink
+{
+    private final Content.Sink _delegate;
+    private final ByteBufferPool _bufferPool;
+    private final boolean _direct;
+    private final int _maxBufferSize;
+    private CountingByteBufferAccumulator _accumulator;
+    private boolean _firstWrite = true;
+    private boolean _lastWritten;
+
+    public BufferedContentSink(Content.Sink delegate, ByteBufferPool bufferPool, boolean direct, int maxBufferSize)
+    {
+        if (maxBufferSize <= 0)
+            throw new IllegalArgumentException("maxBufferSize must be > 0, was: " + maxBufferSize);
+        _delegate = delegate;
+        _bufferPool = (bufferPool == null) ? new ByteBufferPool.NonPooling() : bufferPool;
+        _direct = direct;
+        _maxBufferSize = maxBufferSize;
+    }
+
+    @Override
+    public void write(boolean last, ByteBuffer byteBuffer, Callback callback)
+    {
+        if (_lastWritten)
+        {
+            callback.failed(new IOException("complete"));
+            return;
+        }
+        if (_firstWrite)
+        {
+            _accumulator = new CountingByteBufferAccumulator(_bufferPool, _direct, _maxBufferSize);
+            _firstWrite = false;
+        }
+        _lastWritten |= last;
+
+        ByteBuffer current = byteBuffer != null ? byteBuffer : BufferUtil.EMPTY_BUFFER;
+        IteratingNestedCallback writer = new IteratingNestedCallback(callback)
+        {
+            private boolean complete;
+
+            @Override
+            protected Action process()
```

Comment:
I really don't think we need to iterate inside write.   There are only a few cases:
1. small write that fits in the buffer is aggregated
2. small write that overflows the aggregate buffer, the buffer is topped up and flushed, remainder is put into aggregate.
3. large write that doesn't fit in buffer, flushed any aggregate, then write buffer directly to delegate without copying.

So mostly there is only a single write, and at worse there are 2 writes.   All can be handled with special purpose callback wrappers rather than iteration. 

Our main concern for special callback is stack over flow from recursion prevention.